src/async.rs

```diff
@@ -269,3 +275,295 @@ impl ConnectionLike for Connection {
+                }
+            }
+        };
+        match sender.send(item) {
+            Ok(_) => (),
+            Err(_) => {
+                // `Err` means that the receiver were dropped in which case it does not
+                // care about the output and we can continue by just dropping the value
+                // and sender
+            }
+        }
+    }
+
+    fn poll_write(&mut self) -> Poll<Option<()>, T::Error> {
+        loop {
+            match mem::replace(&mut self.send_state, SendState::Free) {
+                SendState::Unsent(item) => match self.sink_stream.start_send(item)? {
+                    AsyncSink::Ready => {
+                        // Try to take another incoming and start sending it as well
+                        // before completing the send to give the sink a chance to coalesce
+                        // multiple messages into a large one.
+                        match self.incoming.poll() {
+                            Ok(Async::NotReady) | Ok(Async::Ready(None)) | Err(()) => {
+                                self.send_state = SendState::Sent
+                            }
+                            Ok(Async::Ready(Some((item, senders)))) => {
+                                self.in_flight.push_back(senders);
+                                self.send_state = SendState::Unsent(item);
+                            }
+                        }
+                    }
+                    AsyncSink::NotReady(item) => {
+                        self.send_state = SendState::Unsent(item);
+                        return Ok(Async::NotReady);
+                    }
+                },
+                SendState::Sent => {
+                    match self.sink_stream.poll_complete()? {
+                        Async::Ready(()) => (),
+                        Async::NotReady => {
+                            self.send_state = SendState::Sent;
+                            return Ok(Async::NotReady);
+                        }
+                    }
+                    self.send_state = SendState::Free;
+                    return Ok(Async::Ready(Some(())));
+                }
+                SendState::Free => {
+                    let (item, senders) = match self.incoming.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(Some(x))) => x,
+                        Ok(Async::Ready(None)) | Err(()) => return Ok(Async::Ready(None)),
+                    };
+                    self.in_flight.push_back(senders);
+                    self.send_state = SendState::Unsent(item);
+                }
+            }
+        }
+    }
+}
+
+impl<T> Pipeline<T>
+where
+    T: Sink<SinkError = <T as Stream>::Error> + Stream + Send + 'static,
+    T::SinkItem: Send,
+    T::Item: Send,
+    T::Error: Send,
+    T::Error: ::std::fmt::Debug,
+{
+    fn new(sink_stream: T) -> Self {
+        let (sender, receiver) = mpsc::unbounded();
```

Comment:
Fixed, 50 was chosen by a fair dice roll. I think in the long term this could be implemented with https://github.com/tower-rs/tower and we could get this (and other niceties) for free.